File: app/lib/models/cve_severity.dart
import 'package:flutter/material.dart';

/// Display-time normalization of the two severity vocabularies that
/// coexist in `articles.tags`.
///
/// The scraper deliberately does NOT merge them (see CLAUDE.md, "Tag
/// conventions"): Red Hat's Hydra API says `low/moderate/important/
/// critical`, NVD says `low/medium/high/critical`, and they are genuinely
/// different scales — Red Hat's `important` spans NVD's `high` *and*
/// `critical`. `cve_enrichment.py` records which source won rather than
/// mapping between them, so both vocabularies land in the same column.
///
/// That's correct for storage and useless for a UI: a severity filter
/// can't offer the user six overlapping buckets. So we collapse to ONE
/// scale **at display time only** — nothing here writes back, and the
/// stored tags stay exactly as the scraper wrote them.
///
/// The collapse is lossy in the direction Red Hat's scale is coarser
/// (`important` → [high] loses the fact that it might have been an NVD
/// `critical`). That's an accepted tradeoff for a filterable UI; anything
/// needing the true source scale must read the tags directly.
enum CveSeverity {
  critical,
  high,
  medium,
  low;

  /// Rank for sorting — higher is more severe.
  int get rank => switch (this) {
        CveSeverity.critical => 4,
        CveSeverity.high => 3,
        CveSeverity.medium => 2,
        CveSeverity.low => 1,
      };

  /// Uppercase badge text.
  String get label => name.toUpperCase();

  /// Badge accent. Every bucket gets one — unlike `article_card.dart`,
  /// which styles only critical/important/moderate and lets NVD-sourced
  /// `high`/`medium` fall through to a plain SECURITY badge.
  Color get color => switch (this) {
        CveSeverity.critical => const Color(0xFFFF0000),
        CveSeverity.high => const Color(0xFFFF6600),
        CveSeverity.medium => const Color(0xFFFFAA00),
        CveSeverity.low => const Color(0xFF888888),
      };

  /// Maps a single raw severity word from either vocabulary onto the
  /// display scale. Returns null for anything unrecognised, which is how
  /// [fromTags] skips the ~dozen non-severity tags on a CVE article.
  static CveSeverity? fromWord(String raw) {
    switch (raw.toLowerCase().trim()) {
      case 'critical':
        return CveSeverity.critical;
      // Red Hat `important` and NVD `high` collapse to the same bucket.
      case 'important':
      case 'high':
        return CveSeverity.high;
      // Red Hat `moderate` and NVD `medium` likewise.
      case 'moderate':
      case 'medium':
        return CveSeverity.medium;
      case 'low':
        return CveSeverity.low;
      default:
        return null;
    }
  }

  /// The article's severity: the most severe recognised word in [tags].
  ///
  /// Takes the max rather than the first match because a multi-CVE
  /// article can carry more than one severity word, and the scraper
  /// already stamps such articles with the *max* CVSS score — taking the
  /// max here keeps the badge and the score describing the same CVE.
  static CveSeverity? fromTags(Iterable<String> tags) {
    CveSeverity? worst;
    for (final tag in tags) {
      final s = fromWord(tag);
      if (s == null) continue;
      if (worst == null || s.rank > worst.rank) worst = s;
    }
    return worst;
  }
}

/// Parses the `cvss:X.X` tag into a score.
///
/// The scraper's contract is exactly one `cvss:` tag per article (see
/// CLAUDE.md) — but that invariant is enforced on the write side, and a
/// regression there would silently make this vary between renders. So we
/// take the max of whatever we find: deterministic regardless, and it
/// matches [CveSeverity.fromTags] picking the worst severity.
double? cvssFromTags(Iterable<String> tags) {
  double? best;
  for (final tag in tags) {
    final lower = tag.toLowerCase();
    if (!lower.startsWith('cvss:')) continue;
    final value = double.tryParse(lower.substring(5).trim());
    if (value == null) continue;
    if (best == null || value > best) best = value;
  }
  return best;
}

/// Every `CVE-YYYY-N` id on the article, uppercased, in tag order.
///
/// The scraper writes these uppercase; we match case-insensitively and
/// normalize because the id is rendered verbatim.
List<String> cveIdsFromTags(Iterable<String> tags) {
  final ids = <String>[];
  for (final tag in tags) {
    if (!tag.toLowerCase().startsWith('cve-')) continue;
    final id = tag.toUpperCase();
    if (!ids.contains(id)) ids.add(id);
  }
  return ids;
}

File: scraper/sources/cve_severity.py
"""
Routes a CVE-tagged article to its per-severity FCM topic.

**This is one half of a paired contract.** The other half is
``app/lib/models/cve_severity.dart`` (``CveSeverity.fromWord`` /
``fromTags``), which collapses the same two vocabularies for display on the
CVE screen. The two MUST agree: a user filters the CVE screen to "High" and
subscribes to High notifications expecting the same set of CVEs.
``tests/test_cve_severity.py`` parses the Dart source and fails if either
side is edited alone. Change both together.

Why the map is not six inline string comparisons
------------------------------------------------
Severity arrives in ``articles.tags`` in TWO vocabularies that the scraper
deliberately never merges (see CLAUDE.md, "Tag conventions"): Red Hat's
Hydra API says ``low/moderate/important/critical``, NVD says
``low/medium/high/critical``. ``cve_enrichment.py`` records which source won
rather than mapping between them, so both land in the same column.

The consequence is a silent-failure trap. Routing ``cve_high`` on the
literal ``high`` tag looks correct and is catastrophically wrong: measured
against the live table (331 CVE-tagged articles), Red Hat's ``important``
accounts for 194 of them and NVD's ``high`` for 11 — a ~15:1 split. Dropping
the Red Hat words would silently discard 305 of 331 articles (92%) while the
system continued to look healthy: notifications would still fire, tests
asserting "a high CVE routes to cve_high" would still pass, and only the
Red Hat-sourced majority would vanish.

So the mapping is an explicit table, keyed by every raw word we accept, with
the vocabulary each word comes from named in a comment.
"""

from scraper.models import Article

# Raw severity word (either vocabulary) -> FCM topic.
#
# DO NOT "simplify" this by dropping what look like synonyms. `important`
# and `moderate` are Red Hat's words and carry the overwhelming majority of
# real traffic; `high` and `medium` are NVD's. Removing either pair silently
# stops notifications for that source. See the module docstring.
SEVERITY_TOPICS: dict[str, str] = {
    "critical": "cve_critical",   # both vocabularies
    "important": "cve_high",      # Red Hat — Red Hat's "important" IS High
    "high": "cve_high",           # NVD
    "moderate": "cve_medium",     # Red Hat
    "medium": "cve_medium",       # NVD
    "low": "cve_low",             # both vocabularies
}

# Severity ordering, mirroring `CveSeverity.rank` in cve_severity.dart.
# Used to pick the worst severity on a multi-CVE article.
SEVERITY_RANK: dict[str, int] = {
    "cve_critical": 4,
    "cve_high": 3,
    "cve_medium": 2,
    "cve_low": 1,
}

# Every topic this module can route to, worst first. The client subscribes to
# these same strings (`kCveTopics` in notification_service.dart).
CVE_TOPICS: list[str] = ["cve_critical", "cve_high", "cve_medium", "cve_low"]


def severity_word_for_article(article: Article) -> str | None:
    """The worst recognised severity word on the article, or None.

    Takes the max rather than the first match, mirroring Dart's
    ``CveSeverity.fromTags``: a multi-CVE article can carry more than one
    severity word, and the scraper already stamps such articles with the
    *max* CVSS score — taking the max here keeps the notification and the
    score describing the same CVE.
    """
    worst_word: str | None = None
    worst_rank = 0
    for tag in article.tags or []:
        topic = SEVERITY_TOPICS.get(tag.lower().strip())
        if topic is None:
            continue
        rank = SEVERITY_RANK[topic]
        if rank > worst_rank:
            worst_rank = rank
            worst_word = tag.lower().strip()
    return worst_word


def topic_for_article(article: Article) -> str | None:
    """The FCM topic this article should push to, or None if unroutable.

    None means "no recognised severity word in tags" — either the article
    carries no severity at all (2 of 331 live articles) or it carries a word
    outside both vocabularies. The caller MUST log and skip rather than
    guessing a bucket: a wrong guess either wakes people for a low, or
    silently swallows a critical.
    """
    word = severity_word_for_article(article)
    return SEVERITY_TOPICS[word] if word else None
